src/lib/include/simd/simd_float_datastructures.hpp

```diff
@@ -0,0 +1,178 @@
+#ifndef DISTOPIA_SIMD_FLOAT_DATASTRUCTURES_H
+#define DISTOPIA_SIMD_FLOAT_DATASTRUCTURES_H
+
+#include "simd_config.h"
+#include "simd_doubles.h"
+#include "simd_floats.h"
+#include <immintrin.h>
+
+// 3x8 structure of packed floats
+class SimdFloatX8X3 {
+public:
+  SimdFloatX8X3(SimdFloatX8 a, SimdFloatX8 b, SimdFloatX8 c);
+  inline void JNG_AOS2SOASwizzle();
+  inline void AOS2SOASwizzle();
+  inline void SOA2AOSSwizzle();
+  SimdFloatX8 x;
+  SimdFloatX8 y;
+  SimdFloatX8 z;
+};
+
+SimdFloatX8X3::SimdFloatX8X3(SimdFloatX8 a, SimdFloatX8 b, SimdFloatX8 c)
+    : x(a), y(b), z(c) {}
+
+/* performs an in place transpose of the data in a, b and c from
+SOA to AOS form.   x,y,z,x,y,z.. to  x0-7,y0-7,z0-7  */
+inline void SimdFloatX8X3::AOS2SOASwizzle() {
+  /* based on Intel 3D vector norm with intrinsics */
+
+  /* x = x0y0z0x1y1z1x2y2 */
+  /* y = z2x3y3z3x4y4z4x5 */
+  /* z = y6z6x7y7z7x8y8z8 */
+
+  /* break into smaller chunks to swizzle, this may be slower than JNG
+    implementation */
+  __m128 m0 = _mm256_extractf128_ps(x.contents, 0);
+  __m128 m1 = _mm256_extractf128_ps(x.contents, 1);
+  __m128 m2 = _mm256_extractf128_ps(y.contents, 0);
+  __m128 m3 = _mm256_extractf128_ps(y.contents, 1);
+  __m128 m4 = _mm256_extractf128_ps(z.contents, 0);
+  __m128 m5 = _mm256_extractf128_ps(z.contents, 1);
+
+  // load lower halves upper is undefined
+  __m256 m03 = _mm256_castps128_ps256(m0);
+  __m256 m14 = _mm256_castps128_ps256(m1);
+  __m256 m25 = _mm256_castps128_ps256(m2);
+  m03 = _mm256_insertf128_ps(m03, m3, 1); // load upper halves
+  m14 = _mm256_insertf128_ps(m14, m4, 1);
```

Comment:
Use [`_mm256_permute2f128_ps`](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#techs=AVX&text=_mm256_permute2f128_ps&expand=5189,5189,622,4172) instead of `_mm256_extractf128_ps` and `_mm256_insertf128_ps`: it’s half the cost.

Reply:
👍